tests/cpp/test_multidevice_overlap.cpp

```diff
@@ -0,0 +1,242 @@
+  std::string indent = "  ";
+  out << "params:{\n"
+      << indent << "backend_type=" << params.backend_type << "\n"
+      << indent << "M=" << params.M << "\n"
+      << indent << "K=" << params.K << "\n"
+      << indent << "N=" << params.N << "\n"
+      << indent << "S=" << params.S << "\n"
+      << indent << "use_different_streams=" << params.use_different_streams
+      << "\n"
+      << "}";
+  return out;
+}
+
+class OverlapTest : public MultiDeviceTest {
+ protected:
+  OverlapTestParams params;
+
+  int64_t num_devices_;
+  int64_t my_device_index_;
+  at::Tensor ta_, tb_, tc_locally_reduced_, tc_, tc_expected_;
+  // stores the backend
+  c10d::Backend* world_communicator_;
+
+  void SetUp() {
+    MultiDeviceTest::SetUp();
+
+    num_devices_ = communicator->size();
+    my_device_index_ = communicator->deviceId();
+    ASSERT_EQ(params.M % (params.S * num_devices_), 0);
+    ASSERT_EQ(params.K % num_devices_, 0);
+
+    // Setup the world communicators
+    std::vector<int64_t> devices(num_devices_);
+    std::iota(devices.begin(), devices.end(), 0);
+    world_communicator_ =
+        communicator->getBackendForTeam(devices, params.backend_type);
+
+    // Define I/O and intermediate Tensor shapes
+    // clang-format off
+        std::vector<int64_t> ta_unsharded_sizes       = {params.S, num_devices_, params.M/params.S, params.K/num_devices_, 1       };
+        std::vector<int64_t> ta_sizes                 = {params.S, 1           , params.M/params.S, params.K/num_devices_, 1       };
+        std::vector<int64_t> tb_unsharded_sizes       = {1       , num_devices_, 1                , params.K/num_devices_, params.N};
+        std::vector<int64_t> tb_sizes                 = {1       , 1           , 1                , params.K/num_devices_, params.N};
+        std::vector<int64_t> tc_locally_reduced_sizes = {params.S, 1           , params.M/params.S,                        params.N};
+        std::vector<int64_t> tc_sizes                 = {params.S, 1           , params.M/(params.S*num_devices_)        , params.N};
+    // clang-format on
+
+    // Set up input tensors. We create the full unsharded tensors and define the
+    // actual input as the shard corresponding to the current device. Having the
+    // full unsharded input on each rank makes it possible to compute the
+    // expected result locally, hence, this way of doing is convenient for
+    // validating data correctness.
+    at::TensorOptions options =
+        at::TensorOptions().dtype(at::kFloat).device(communicator->device());
+    auto ta_unsharded = at::randn(ta_unsharded_sizes, options);
+    auto tb_unsharded = at::randn(tb_unsharded_sizes, options);
+    ta_ = at::empty(ta_sizes, options);
+    ta_.copy_(ta_unsharded.slice(1, my_device_index_, my_device_index_ + 1));
+    tb_ = at::empty(tb_sizes, options);
+    tb_.copy_(tb_unsharded.slice(1, my_device_index_, my_device_index_ + 1));
```

Comment:
Nit: could use getSlice to simplify the code a little bit.

Reply:
getSlice operates on axis 0, that's why I'm not using it here. But let me add an "axis" argument to getSlice, so I can use it everywhere